code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,439 @@
+GLOBAL_LIST_INIT(medium_wave, list(/obj/structure/crystal_portal/small=4, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=1)) //8
+GLOBAL_LIST_INIT(big_wave, list(/obj/structure/crystal_portal/small=5, /obj/structure/crystal_portal/medium=3, /obj/structure/crystal_portal/big=2, /obj/structure/crystal_portal/huge=1)) //11
+GLOBAL_LIST_INIT(huge_wave, list(/obj/structure/crystal_portal/small=7, /obj/structure/crystal_portal/medium=5, /obj/structure/crystal_portal/big=3, /obj/structure/crystal_portal/huge=2)) //17
+GLOBAL_LIST_EMPTY(crystal_portals)
+/*
+This section is for the event controller
+*/
+/datum/round_event_control/crystal_invasion
+	name = "Crystal Invasion : Small"
+	typepath = /datum/round_event/crystal_invasion
+	weight = 4
+	min_players = 15
+	max_occurrences = 3
+	earliest_start = 25 MINUTES
+
+/datum/round_event/crystal_invasion
+	announceWhen = 1
+	///Stores the type of the wave for the event
+	var/list/wave_type
+	///Is the name of the wave, used to check wich wave will be generated
+	var/wave_name = "small wave"
+	///Max number of portals that can spawn per type of wave
+	var/portal_numbers = 5
+
+/datum/round_event/crystal_invasion/announce()
+	priority_announce("WARNING - Destabilization of the Supermatter Crystal Matrix detected, please stand by waiting further instructions", "Alert", 'sound/misc/notice1.ogg')
+
+/datum/round_event/crystal_invasion/New()
+	. = ..()
+	if(!wave_type)
+		choose_wave_type()
+
+///Choose the type of the wave
+/datum/round_event/crystal_invasion/proc/choose_wave_type()
+	if(!wave_name)
+		wave_name = pickweight(list(
+			"small wave" = 50,
+			"medium wave" = 35,
+			"big wave" = 10,
+			"huge wave" = 5))
+	switch(wave_name)
+		if("small wave")
+			wave_type = GLOB.small_wave
+		if("medium wave")
+			wave_type = GLOB.medium_wave
+		if("big wave")
+			wave_type = GLOB.big_wave
+		if("huge wave")
+			wave_type = GLOB.huge_wave
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+	sleep(10 SECONDS)
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
+			WARNING("No engine found.")
+			kill()
+		if(istype(temp, /obj/machinery/power/supermatter_crystal/shard))
+			message_admins("found a shard, skipping it")
+			continue
+		sm_crystal += temp
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
+	crystal.destabilize()
+
+	priority_announce("WARNING - Detected numerous energy fluctuation generated from your Supermatter; we estimate a [wave_name] of crystal-like creatures \
+						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Collect the \[REDACTED] \
+						anomaly from the remains of the portals and use it in a crystal stabilizer to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario", "Alert", 'sound/misc/notice1.ogg')
```

Comment:
```suggestion
	priority_announce("WARNING - Numerous energy fluctuations have been detected from your Supermatter; we estimate a [wave_name] of crystalline creatures \
						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Harvest a \[REDACTED] \
						anomaly from a portal, place it inside a crystal stabilizer, and inject it into your Supermatter to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario from occurring.", "Alert", 'sound/misc/notice1.ogg')
```
also wtf is a ZK-Lambda-Class Cosmic Fragmentation Scenario

Reply:
Thanks for the corrections 

is a SCP reference